src/WindowsUtils/Injection.cpp

```diff
@@ -0,0 +1,274 @@
+        absl::StrFormat("Multiple modules with the name \"%s\" found", module_name));
+  return ErrorMessage("Could not find module in target process");
+}
+
+ErrorMessageOr<Module> FindModule(uint32_t pid, std::string_view module_name,
+                                  uint32_t num_retries) {
+  while (true) {
+    ErrorMessageOr<Module> result = FindModule(pid, module_name);
+    if (result.has_value() || num_retries-- <= 0) return result;
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  }
+}
+
+ErrorMessageOr<void> EnsureModuleIsNotAlreadyLoaded(uint32_t pid, std::string_view module_name) {
+  ErrorMessageOr<Module> module = FindModule(pid, module_name);
+  if (module.has_value()) {
+    return ErrorMessage(
+        absl::StrFormat("Module %s is already loaded in process %u", module_name, pid));
+  }
+  return outcome::success();
+}
+
+}  // namespace
+
+ErrorMessageOr<void> InjectDll(uint32_t pid, std::filesystem::path dll_path) {
+  const std::string dll_name = dll_path.string();
+  SCOPED_TIMED_LOG("Injecting dll \"%s\" in process %u", dll_name, pid);
+
+  OUTCOME_TRY(ValidatePath(dll_path));
+  OUTCOME_TRY(EnsureModuleIsNotAlreadyLoaded(pid, dll_path.filename().string()));
+
+  // Inject dll by calling "LoadLibraryA" in remote process with the name of our dll as parameter.
+  OUTCOME_TRY(CreateRemoteThread(pid, "kernel32.dll", "LoadLibraryA", ToByteBuffer(dll_name)));
+
+  // Find injected dll in target process. Allow for retries as the loading might take some time.
+  OUTCOME_TRY(Module module, FindModule(pid, dll_path.filename().string(), /*num_retries=*/10));
+
+  LOG("Module %s successfully injected in process %u", dll_name, pid);
+  return outcome::success();
+}
+
+ErrorMessageOr<void> CreateRemoteThread(uint32_t pid, std::string_view module_name,
+                                        std::string_view function_name,
+                                        std::vector<uint8_t> parameter) {
+  OUTCOME_TRY(uint64_t function_address, GetRemoteProcAddress(pid, module_name, function_name));
+  OUTCOME_TRY(HANDLE handle, GetHandleFromPid(pid));
+  orbit_base::unique_resource handle_closer(handle, ::CloseHandle);
+
+  // Write parameter to remote process memory.
+  void* parameter_address = nullptr;
+  if (!parameter.empty()) {
+    OUTCOME_TRY(void* address, RemoteWrite(handle, parameter));
+    parameter_address = address;
+  }
+
+  if (!::CreateRemoteThread(handle, /*lpThreadAttributes=*/0, /*dwStackSize=*/0,
+                            absl::bit_cast<LPTHREAD_START_ROUTINE>(function_address),
+                            parameter_address,
+                            /*dwCreationFlags=*/0, /*lpThreadId=*/0)) {
+    return ErrorMessage(absl::StrFormat("CreateRemoteThread failed with error %s",
+                                        orbit_base::GetLastErrorAsString()));
+  }
+
+  return outcome::success();
+}
+
+ErrorMessageOr<uint64_t> GetRemoteProcAddress(uint32_t pid, std::string_view module_name,
```

Comment:
I think that this functions really needs some documenting comments. For example explaining some terminology. It's very hard to follow.